api/odl.py

```diff
@@ -0,0 +1,598 @@
+        except ValueError, e:
+            raise BadResponseException(url, "License Status Document was not valid JSON.")
+        if status_doc.get("status") not in self.STATUS_VALUES:
+            raise BadResponseException(url, "License Status Document had an unknown status value.")
+        return status_doc
+
+    def checkin(self, patron, pin, licensepool):
+        """Return a loan early."""
+        _db = Session.object_session(patron)
+
+        loan = _db.query(Loan).filter(
+            Loan.patron==patron
+        ).filter(
+            Loan.license_pool_id==licensepool.id
+        )
+        if loan.count() < 1:
+            raise NotCheckedOut()
+        loan = loan.one()
+
+        doc = self.get_license_status_document(licensepool, patron, loan)
+        status = doc.get("status")
+        if status in [self.REVOKED_STATUS, self.RETURNED_STATUS, self.CANCELLED_STATUS, self.EXPIRED_STATUS]:
+            # This loan was already returned early or revoked by the distributor, or it expired.
+            raise NotCheckedOut()
+        elif status == self.ACTIVE_STATUS:
+            # This loan has already been fulfilled, so it needs to be returned through the DRM system.
+            raise CannotReturn()
+
+        return_url = doc.get("links", {}).get("return", {}).get("href")
+        if return_url:
+            # Hit the distributor's return link.
+            self._get(return_url)
+            # Get the status document again to make sure the return was successful,
+            # and if so update the pool availability and delete the local loan.
+            self.update_loan(loan)
+        else:
+            # The distributor didn't provide a link to return this loan.
+            raise CannotReturn()
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """Create a new loan."""
+        if licensepool.licenses_owned < 1:
+            raise NoLicenses()
+
+        if licensepool.licenses_available < 1:
+            raise NoAvailableCopies()
+
+        _db = Session.object_session(patron)
+        loan = _db.query(Loan).filter(
+            Loan.patron==patron
+        ).filter(
+            Loan.license_pool_id==licensepool.id
+        )
+        if loan.count() > 0:
+            raise AlreadyCheckedOut()
+
+        doc = self.get_license_status_document(licensepool, patron)
+        status = doc.get("status")
+
+        if status not in [self.READY_STATUS, self.ACTIVE_STATUS]:
+            # Something went wrong with this loan and we don't actually
+            # have the book checked out. This should never happen.
+            raise CannotLoan()
+
+        # We have successfully borrowed this book.
+        licensepool.licenses_available -= 1
+
+        expires = doc.get("potential_rights", {}).get("end")
+        expires = datetime.datetime.strptime(expires, self.TIME_FORMAT)
+        external_identifier = doc.get("links", {}).get("self", {}).get("href")
```

Comment:
Given the importance of the external_identifier I think the lack of one should be treated as an error condition.